### DISCUSSION TITLE
preferReadReplica not respected with createQueryBuilder for Postgres

### DISCUSSION THREAD

**Discussion by jtuppy** [mikro-orm/mikro-orm#4847] [General]
First, just want to say I am a huge fan of this library, it is very well done and thank you for the hard work making this for the community.

**Describe the bug**

I have a question about understanding how connection type is being decided and the `preferReadReplica` option. We wanted to gradually start adopting a read replica for certain queries in our app so wanted to leverage the `preferReadReplica: false` to selectively choose which queries would start using it. It appears that this option wasn't respected when using query builder, because we were hitting the read replica when none of our queries should have been using it.

Since we use the PostgresSqlDriver, it seems like the connectionType is not specified when using createQueryBuilder from the EntityManager.

It seems the createQueryBuilder function in AbstractSqlDriver respects these options when calling `resolveConnectionType`
https://github.com/mikro-orm/mikro-orm/blob/5a24ceda3aa548447077b3d7dfc2c3c6e6484f62/packages/knex/src/AbstractSqlDriver.ts#L871-L880

but when calling it from the SqlEntityManager, the connectionType remains undefined
https://github.com/mikro-orm/mikro-orm/blob/5a24ceda3aa548447077b3d7dfc2c3c6e6484f62/packages/knex/src/SqlEntityManager.ts#L15-L19

which means it could can take on a read connection in `execute`
https://github.com/mikro-orm/mikro-orm/blob/5a24ceda3aa548447077b3d7dfc2c3c6e6484f62/packages/knex/src/query/QueryBuilder.ts#L643

Happy to put up a fix if this is indeed an issue, but wanted to make sure I'm tracing this out correctly first or if there is something I missed.

**To Reproduce**
1. Connect to postgres
2. set `preferReadReplica` to false
3. em.createQueryBuilder(Entity).select('...').execute() seems to use the read replica

**Expected behavior**
Would expect that the preferReadReplica is respected using the query builder

**Versions**

| Dependency | Version |
| - | - |
| node | 20 |
| mikro-orm | 5.6.16 |
| your-driver | Postgres |

**Comment by B4nan** [comment]
I don't recall when we added the `preferReadReplicas` option, I think this was a community addition. Looking at the code you are probably right, it should be respected in the QB too, same way as in the driver class.

PR welcome 👍 It will be a bit more work than just using the `resolveConnectionType` method (feel free to make it public if needed, just add `@internal` annotation to it), direct QB usage does not necessarily have the transaction context, and we need to use write connections for write queries based on the QB type.

**Reply by B4nan** [reply]
FYI https://github.com/mikro-orm/mikro-orm/commit/22e140e986420105c1b5941aae92a7bca1be6fef?w=1

**Reply by jtuppy** [reply]
Wow, coming back to this, and seems done, thank you!
